lib/src/fty_nut_server.cc

```diff
@@ -68,103 +88,174 @@ static bool get_initial_licensing(StateManager::Writer& state_writer, mlm_client
+            if (r < 0) {
+                log_warning("Sending ASSET_DETAIL message for %s failed", asset.get());
+            }
+            else {
+                uuids.emplace(uuid_str); // ok, remember uuid
+            }
+
+            asset = zmsg_popstr(reply); // next
         }
-        asset = zmsg_popstr(reply);
+        //log_debug("%zu ASSET_DETAIL requests sent", uuids.size());
     }
+
+    log_debug("recv %zu ASSET_DETAIL responses", uuids.size());
     bool changed = false;
-    while (!uuids.empty()) {
-        zmsg_t* reply1 = mlm_client_recv(client);
-        if (uuids.erase(ZstrGuard(zmsg_popstr(reply1)).get()) == 0) {
-            log_warning("Mismatching response to an ASSET_DETAIL request");
-            zmsg_destroy(&reply1);
-            continue;
+    {
+        size_t noResponseCnt = 0;
+        while (uuids.size() > noResponseCnt) {
+            if (zsys_interrupted) {
+                return; //$TERM
+            }
+
+            // recv a response
+            zmsg_t* msg = NULL;
+            if (!zpoller_wait(poller, recvTimeout)) {
+                log_warning("Getting ASSET_DETAIL response timed out (%d ms)", recvTimeout);
+            }
+            else {
+                msg = mlm_client_recv(client);
+            }
+            if (!msg) {
+                noResponseCnt++;
+                continue;
+            }
+
+            char* uuid = zmsg_popstr(msg);
+
+            if (!uuid || uuids.erase(uuid) == 0) {
+                log_warning("Mismatching response to an ASSET_DETAIL request");
+            }
+            else if (!fty_proto_is(msg)) {
+                log_warning("Response to an ASSET_DETAIL message is not fty_proto");
+            }
+            else if (state_writer.getState().updateFromMsg(&msg)) {
+                changed = true;
+            }
+
+            zstr_free(&uuid);
+            zmsg_destroy(&msg);
         }
-        if (!fty_proto_is(reply1)) {
-            log_warning("Response to an ASSET_DETAIL message is not fty_proto");
-            zmsg_destroy(&reply1);
-            continue;
+
+        if (uuids.size() != 0) {
+            log_warning("Missed %zu ASSET_DETAIL responses", uuids.size());
         }
-        if (state_writer.getState().updateFromMsg(&reply1))
-            changed = true;
-        zmsg_destroy(&reply1); // secure
     }
+
     if (query_licensing) {
-        if (get_initial_licensing(state_writer, client))
+        if (get_initial_licensing(state_writer, client)) {
             changed = true;
+        }
     }
-    if (changed)
-        state_writer.commit();
-    log_info("Initial ASSETS request complete (%zd/%zd powerdevices, %zd/%zd sensors)",
-        state_writer.getState().getPowerDevices().size(), state_writer.getState().getAllPowerDevices().size(),
-        state_writer.getState().getSensors().size(), state_writer.getState().getAllSensors().size());
-}
-
-uint64_t polling_timeout(uint64_t last_poll, uint64_t polling_timeout)
```

Comment:
Protection for many devices not needed ?

Reply:
Yes, i changed that... the previous effect was to focus on polling action, loosing reactivity on notification messages processing.
Anyway, the case is rarely reached, and don't changed poll/message processing even n our endurance setup...